### DISCUSSION TITLE
How to checkout private submodules from inside of codespaces

### DISCUSSION THREAD

**Discussion by crazyhappygame** [community/community#51011] [Codespaces] [answered]
### Select Topic Area

Question

### Body

For my account I have private repository with private submodule from the same account.
How can I checkout private submodules?
```
$ git submodule init
....OK
$ git submodule
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: clone of 'git@github.com:crazyhappygame/xxxxx.git' into submodule path '/workspaces/sub/external/xxxxx' failed
```

**Comment by pie972** [comment]
First, you need to have read access to the repository and its submodules. Then, you need to setup authentication (I can help you with this step, in case you dont know how) to access the private repository and its submodules. Finally, you can initialize and updates the submodules
```
$ git submodule init
....OK
$ git submodule update
Cloning into '/workspaces/sub/external/xxxxx'...
Enter passphrase for key '/home/codespace/.ssh/id_rsa': 
remote: Enumerating objects: 16, done.
remote: Counting objects: 100% (16/16), done.
remote: Compressing objects: 100% (12/12), done.
remote: Total 16 (delta 1), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (16/16), done.
Resolving deltas: 100% (1/1), done.
```

**Comment by crazyhappygame** [comment]
Thank you for answering. How can I setup codespace authentification, permissions?

I have access from codespaces to the private repository for which the codespace was created. The submodule it is my private repository. I can not checkout submodule from codespace

**Reply by pie972** [reply]
1. Create a personal access token (PAT) with the necessary permissions. You can create a token in the "Settings" section of your GitHub account. Make sure the token has the "repo" and "workflow" scopes, as these are required for Codespaces.
2. Open the Codespaces settings by clicking on the gear icon in the bottom left corner of the Codespaces window.
3. Click on "Advanced".
4. Under "Environment Variables", click on "Add Variable".
5. Add a new environment variable with the name "GITHUB_TOKEN" and the value of the PAT you created in step 1.
6. Save the settings and restart the Codespace.

Once you have set up authentication, you should be able to access your private repositories and submodules from within the Codespace.

**Reply by crazyhappygame** [reply]
Thank you I created PAT (personal access token) with "repo" and "workflow" scopes.
Unfortunately, I am not sure how to add env variable with steps 2,3,4,5. Could you help me?

I set GITHUB_TOKEN from terminal using PAT
```
export GITHUB_TOKEN=<PAT>
```
But still have
```
$ git submodule update
Cloning into '/workspaces/sub/external/xxxx'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: clone of 'git@github.com:crazyhappygame/xxxx.git' into submodule path '/workspaces/sub/external/xxxx' failed
```

**Reply by pie972** [reply]
From what you wrote here, it looks like you have set the Github_Token envirn. correctly but it seems like its not used by Git to authenticate with Github, so can you please check ur .gitmodules file of ur main repo and include the token in the url of the sobmodules
It should look like something like the following
```
[submodule "external/xxxx"]
    path = external/xxxx
    url = https://x-access-token:${GITHUB_TOKEN}@github.com/crazyhappygame/xxxx.git
```
Then update the submodule using 
```
git submodule update --init --recursive
```
Finally, you can verify if you can access the submodule by navigating to its directory and running 'git pull' or any other Git command

**Comment by crazyhappygame** [comment]
FYI I use git@github.com instead https in submodules - not sure if it makes difference

**Reply by pie972** [reply]
When you use 'git@github.com' instead of 'https://github.com' in your submodules, Git will try to authenticate using SSH keys instead of HTTPS credentials. So, yes, there is a difference.

**Comment by crazyhappygame** [comment]
I tried to follow https://docs.github.com/en/codespaces/managing-your-codespaces/managing-repository-access-for-your-codespaces
But it did not work for me as well ..

**Comment by crazyhappygame** [comment, marked as answer]
@pie972 What I need I think is to deploy ssh key to codespaces. How can I deploy ssh keys to my codespace?

**Reply by pie972** [reply]
1. Generate a new SSH key pair on your Codespace using the following command (if you don't have one already)
```
ssh-keygen -t ed25519 -C "your_email@example.com"
```
You can use different algorithm other than Ed25519 if you want
2. Copy the public key to your GitHub account using the following command
```
cat ~/.ssh/id_ed25519.pub
```
3. Add the private key to your SSH agent using the following
```
ssh-add ~/.ssh/id_ed25519
```

**Reply by crazyhappygame** [reply]
OK. I generated ssh key pair. Added public key to GitHub account. And it works now.

Thank you very much.
The only one question I have is to where to store private key in codespaces? I will need it when I will rebuild image.
For now I added it into secrets. But not sure if it is the most efficient way ....

**Reply by pie972** [reply]
you can store the private key in your codespace's file system, and add the appropriate SSH configuration to your ```~/.ssh/config``` file to ensure that Git can find and use the key.

**Reply by pie972** [reply]
To do this, you can:
1. In your codespace, create a .ssh directory in your home directory (if it doesn't already exist):
```
mkdir ~/.ssh
```
2. Copy your private key file to the .ssh directory
```
scp path/to/private_key user@codespace:/home/user/.ssh/
```
3. Set the correct permissions
i.e. :
```
chmod 600 ~/.ssh/private_key
```
4. Add the following configuration to your ~/.ssh/config file to tell Git to use the private key when connecting to GitHub:
```
Host github.com
  IdentityFile ~/.ssh/private_key
  IdentitiesOnly yes
```

**Reply by pie972** [reply]
Also, if your original question is answered, can you mark it as 'answered' instead of 'Unanswered' ? and we can keep the discussion if you have any further questions. Anyways, Best of luck ^^

**Comment by crazyhappygame** [comment]
Thank you LL. Closing

**Reply by pie972** [reply]
You welcome ^^

**Comment by jessfraz** [comment]
If you follow the instructions here for adding permissions for repos:  https://docs.github.com/en/codespaces/managing-your-codespaces/managing-repository-access-for-your-codespaces be sure to add `content: read` for the specific submodule you repos

Then you can add the following to your devcontainer.json to checkout submodules and rewrite git paths:
```
"updateContentCommand": "git config --global url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf ssh://git@github.com/ && git config --global url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf git@github.com/ && git config --global url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf git@github.com: && git submodule update --init --recursive",
```

**Reply by RIscRIpt** [reply]
Thanks!

I found that `updateContentCommand` is not executed when a codespace container is stopped and started with a new `GITHUB_TOKEN`. Where as `initializeCommand` is executed each time.

Additionally, to make commits and push them I needed write access, I am not sure which exactly, so I used `write-all`.

Finally, above `git config` commands are missing `--add` flag. Without it only the last config value is saved.

```json
{
  "customizations": {
    "codespaces": {
      "repositories": {
        "org/repo": { "permissions": "write-all" }
      }
    }
  },
  "initializeCommand": {
    "git rewrite ssh://git@github.com/": "git config --global --add url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf ssh://git@github.com/",
    "git rewrite git@github.com/": "git config --global --add url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf git@github.com/",
    "git rewrite git@github.com:": "git config --global --add url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf git@github.com:",
    "git clone submodules": "git submodule update --init --recursive"
  }
}
```

`initializeCommand` is configured using [object format](https://containers.dev/implementors/json_reference/#formatting-string-vs-array-properties)

**Reply by jhhhnikt** [reply]
I ended up in this thread after searching for a solution, and came up with the following after looking at your answers @jessfraz and @RIscRIpt .. I think this is okay.

```yaml
{
    // ....
    "postStartCommand": {
        "git and submodules setup": "git config -l | grep -q \"x-access-token:${GITHUB_TOKEN}\" && echo 'Git is already configured ..' || (git config --replace-all --global url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf ssh://git@github.com/ && git config --add --global url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf git@github.com/ && git config --add --global url.https://x-access-token:${GITHUB_TOKEN}@github.com/.insteadOf git@github.com: && git submodule update --init --recursive)"
    },
    "customizations": {
        "codespaces": {
            "repositories": {
                "our-org/our-sub-module-repo": {
                    "permissions": {
                        "contents": "read"
                    }
                }
            }
        }
    }
}
```

Note that I am using `postStartCommand` here to update the configuration if `GITHUB_TOKEN` changes. It will run each time the container is successfully started.

If you split the commands into multiple entry objects, they will run in parallel in that lifecycle step. So instead I run the commands in one entry so they execute in serial.

But thanks for the suggestions to both of you.
